### PR TITLE
test: add unit tests for uploads

### PR DESCRIPTION
**Pull request by thaaaarun**
Fixes #<9>

> It's a good idea to open an issue first for discussion.

- [/ ] Tests pass
- [ ] Appropriate changes to README are included in PR

### CODE REVIEW

File: front_end/test/common.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * Error message returned when a method is called with an invalid filename
+ * @const {string}
+ */
+export const ERR_MSG_INVALID_FILE = `Invalid File Name: 'random_file_name'. File names must be any of "2mib.txt", "64mib.txt" or "256mib.txt"`;
+
+/**
+ * Error message returned when a method is called with an invalid bucket name
+ * @const {string}
+ */
+export const ERR_MSG_INVALID_BUCKET = `Invalid Bucket Name: 'random_bucket_name'. Bucket must be a supported Google Cloud Storage Region Name. View https://cloud.google.com/storage/docs/locations for more information.`;
```
Comment (ddelgrosso1) [inline front_end/test/common.js:27]: Not sure if you are linting tests yet but this line looks long. Probably should be split.

File: front_end/test/uploads.js [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        // uploads.getDurationOfUpload defaults elapsed time to -1 in failure. This is divided by 1000
+        // to convert from millisecs to seconds. -1/1000 = -0.001.
+        it('should return -0.001 if PUT request fails', async () => {
+            sinon.stub(axios, 'put').callsFake(fakeAxiosFailure);
+            sinon.stub(axios, 'get').callsFake(fakeAxiosSuccess);
+
+            const fileName = '2mib.txt';
+            const bucketName = 'us-west1';
+
+            let result = await uploads.getDurationOfUpload(fileName, bucketName);
```
Comment (shaffeeullah) [inline front_end/test/uploads.js:96]: can this be `const`?

File: front_end/test/downloads.js [hunk 3]
```diff
@@ -28,27 +30,24 @@ async function fakeAxiosGet(URL) {
...
-const ERR_MSG_INVALID_FILE = `Invalid File Name: 'random_file_name'. File names must be any of "2mib.txt", "64mib.txt" or "256mib.txt"`;
-const ERR_MSG_INVALID_BUCKET = `Invalid Bucket Name: 'random_bucket_name'. Bucket must be a supported Google Cloud Storage Region Name. View https://cloud.google.com/storage/docs/locations for more information.`;
-
 describe('downloads', () => {
     let downloads;
-    sinon.stub(performance, 'now').callsFake(fakeDateNow);
-    sinon.stub(axios, 'get').callsFake(fakeAxiosGet);
-    
+    sinon.stub(performance, 'now').callsFake(fakePerformanceNow);
+    let spyAxiosGetRequest = sinon.stub(axios, 'get').callsFake(fakeAxiosGet);
```
Comment (shaffeeullah) [inline front_end/test/downloads.js:41]: can this be `const`?

File: front_end/test/downloads.js [hunk 4]
```diff
@@ -103,7 +102,9 @@ describe('downloads', () => {
             const fileName = '2mib.txt';
             const bucketName = 'us-west1';
             await downloads.getDurationInSeconds(fileName, bucketName);
-            assert.deepStrictEqual(downloads._builtURL, `https://storage.googleapis.com/gcsrbpa-us-west1/2mib.txt`);
+
+            let calledWith = spyAxiosGetRequest.args
```
Comment (shaffeeullah) [inline front_end/test/downloads.js:106]: `const`

File: front_end/test/uploads.js [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        uploads = new Uploads();
+    });
+
+    afterEach(() => {
+        sinon.restore();
+    });
+
+    describe('getSignedURL', async () => {
+        it('should form the source of the signed URL correctly', async () => {
+            let spyAxiosGetMethod = sinon.stub(axios, 'get').callsFake(fakeAxiosSuccess);
```
Comment (shaffeeullah) [inline front_end/test/uploads.js:52]: `const`

File: front_end/test/uploads.js [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+            const fileName = '2mib.txt';
+            const bucketName = 'us-west1';
+
+            let result = await uploads.getDurationOfUpload(fileName, bucketName);
+
+            assert.deepStrictEqual(result, -0.001);
+        });
+
+        it('should form axiosOptions correctly', async () => {
+            let spyAxiosPutMethod = sinon.stub(axios, 'put').callsFake(fakeAxiosSuccess);
```
Comment (shaffeeullah) [inline front_end/test/uploads.js:102]: const

File: front_end/test/uploads.js [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+            const bucketName = 'random_bucket_name';
+
+            await assert.rejects(uploads.benchmarkSingleUpload(fileName, bucketName), { message: ERR_MSG_INVALID_BUCKET });
+        });
+
+        it('should return an Array of an Object with bad values if PUT request fails', async () => {
+            sinon.stub(axios, 'get').callsFake(fakeAxiosSuccess);
+            sinon.stub(axios, 'put').callsFake(fakeAxiosFailure);
+
+            let expected = [
```
Comment (shaffeeullah) [inline front_end/test/uploads.js:170]: const

File: front_end/test/uploads.js [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+                    fileSizeBytes: '2097152',
+                    speedBps: '-2097152000.000',
+                    speedMiBps: '-2000.000'
+                }
+            ];
+
+            const fileName = '2mib.txt';
+            const bucketName = 'us-west1';
+
+            let actual = await uploads.benchmarkSingleUpload(fileName, bucketName);
```
Comment (shaffeeullah) [inline front_end/test/uploads.js:185]: const

File: front_end/test/uploads.js [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+            await assert.rejects(uploads.benchmarkSingleUpload(fileName, bucketName), { message: `${ERR_MSG_FAILED_TO_REACH_SERVER}: Error: ${ERR_MSG_AXIOS_FAILURE}` });
+        });
+
+        it('should return an Array of an Object on success', async () => {
+            sinon.stub(axios, 'get').callsFake(fakeAxiosSuccess);
+            sinon.stub(axios, 'put').callsFake(fakeAxiosSuccess);
+            sinon.stub(performance, 'now').callsFake(fakePerformanceNow);
+
+            // SpeedBps and SpeedMiBps are set to 'Infinity' because of the division by 0. 
+            let expected = [
```
Comment (shaffeeullah) [inline front_end/test/uploads.js:206]: const

File: front_end/test/uploads.js [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+                    fileSizeBytes: '2097152',
+                    speedBps: 'Infinity',
+                    speedMiBps: 'Infinity'
+                }
+            ];
+
+            const fileName = '2mib.txt';
+            const bucketName = 'us-west1';
+
+            let actual = await uploads.benchmarkSingleUpload(fileName, bucketName);
```
Comment (shaffeeullah) [inline front_end/test/uploads.js:221]: const

**Review by shaffeeullah — COMMENTED** [review]
A few nit comments about `let` vs `const`. Otherwise, looks good! Your PRs are getting much much cleaner. Awesome job, Tharun!!

File: front_end/test/uploads.js [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+            // NOTE: This test will change during deployment, when signedURL server is not hardcoded to localhost.
+            assert.deepStrictEqual(spyAxiosGetMethod.args[0], [`http://localhost:3000/${bucketName}/${fileName}`]);
+        });
+
+        it('should throw an error when axios request fails', async () => {
+            sinon.stub(axios, 'get').callsFake(fakeAxiosFailure);
+            const fileName = 'random_file_name';
+            const bucketName = 'random_bucket_name';
+
+            await assert.rejects(uploads.getSignedURL(fileName, bucketName), { message: `${ERR_MSG_FAILED_TO_REACH_SERVER}: Error: ${ERR_MSG_AXIOS_FAILURE}` });
```
Comment (ddelgrosso1) [inline front_end/test/uploads.js:67]: Consider making the error its own variable.

**Review by shaffeeullah — APPROVED** [review]
LGTM. Please wait for explicit approval from @ddelgrosso1 before merging.

**Pull request merged**